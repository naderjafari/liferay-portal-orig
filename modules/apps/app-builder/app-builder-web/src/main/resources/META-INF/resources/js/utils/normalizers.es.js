/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {DataConverter} from 'data-engine-taglib';

import {getValidName} from '../utils/utils.es';

export const normalizeNames = ({
	allowEmptyKeys = true,
	defaultName = '',
	localizableValue,
}) => {
	const name = {};

	Object.keys(localizableValue).forEach((languageId) => {
		const value = localizableValue[languageId];
		const normalizedValue = getValidName(defaultName, value)?.trim();

		if (!allowEmptyKeys && !normalizedValue) {
			return;
		}

		name[languageId] = normalizedValue;
	});

	return name;
};

export const normalizeDataDefinition = (dataDefinition) => ({
	...dataDefinition,
	dataDefinitionFields: dataDefinition.dataDefinitionFields.map(
		(dataDefinitionField) => ({
			...dataDefinitionField,

			// Actually showLabel property will be always true
			// because the same property can be controlled by dataLayoutFields

			showLabel: true,
		})
	),
});

export const normalizeDataLayout = ({
	dataDefinition,
	dataLayout,
	dataLayoutBuilder,
	defaultLanguageId,
	editingLanguageId,
}) => {
	const {dataDefinitionFields = []} = dataDefinition;
	const dataLayoutFields = {...dataLayout.dataLayoutFields};

	Object.keys(dataLayoutFields).forEach((field) => {
		if (!dataDefinitionFields.find((item) => item.name === field)) {
			delete dataLayoutFields[field];
		}
	});

	dataDefinitionFields.forEach((dataDefinitionField) => {
		const {
			appContext: [{editingLanguageId}],
			fieldTypes,
		} = dataLayoutBuilder.props;
		const fieldProperties = DataConverter.getDDMSettingsContextWithVisualProperties(
			{
				dataDefinitionField,
				editingLanguageId,
				fieldTypes,
			}
		);

		// Ignore this visual properties because it is treated differently

		delete fieldProperties['label'];
		delete fieldProperties['required'];

		dataLayoutFields[dataDefinitionField.name] = {
			...dataLayoutFields[dataDefinitionField.name],
			...fieldProperties,
		};
	});

	const name = normalizeNames({
		defaultName: Liferay.Language.get('untitled-form-view'),
		localizableValue: dataLayout.name,
	});

	if (!name[defaultLanguageId]) {
		name[defaultLanguageId] = name[editingLanguageId];
	}

	return {
		...dataLayout,
		dataLayoutFields,
		name,
	};
};
