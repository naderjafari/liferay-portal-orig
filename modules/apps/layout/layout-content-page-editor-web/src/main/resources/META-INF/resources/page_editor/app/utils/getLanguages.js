/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {config} from '../../app/config/index';

export default function getLanguages(
	availableLanguages,
	availableSegmentsExperiences,
	segmentsExperienceId
) {
	if (
		!availableSegmentsExperiences ||
		!segmentsExperienceId ||
		!availableSegmentsExperiences[segmentsExperienceId] ||
		segmentsExperienceId === config.defaultSegmentsExperienceId
	) {
		return availableLanguages;
	}

	const languages =
		availableSegmentsExperiences[segmentsExperienceId].languageIds;

	return languages.reduce((acc, languageId) => {
		if (availableLanguages[languageId]) {
			return {
				...acc,
				[languageId]: {...availableLanguages[languageId]},
			};
		}

		return {...acc};
	}, {});
}
