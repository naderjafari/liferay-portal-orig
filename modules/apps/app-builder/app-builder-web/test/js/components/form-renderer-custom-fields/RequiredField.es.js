/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {cleanup, fireEvent, render} from '@testing-library/react';
import React, {createContext} from 'react';

import RequiredField from '../../../../src/main/resources/META-INF/resources/js/components/form-renderer-custom-fields/RequiredField.es';
import * as utils from '../../../../src/main/resources/META-INF/resources/js/components/form-renderer-custom-fields/shared/utils.es';
import {FORM_VIEW} from '../../constants.es';

const DEFAULT_STATE = {
	editingLanguageId: 'en_US',
	focusedCustomObjectField: {},
	focusedField: {
		fieldName: 'RequiredName',
	},
};

const DATA_DEFINITION_FIELD = (label) => ({
	label: {
		en_US: label,
	},
	name: `Required${label}`,
});

const STRUCTURE_LEVEL_STATE = {
	...DEFAULT_STATE,
	dataDefinitionField: {
		...DATA_DEFINITION_FIELD('Name'),
		required: true,
	},
	dataDefinitionFields: [
		{
			...DATA_DEFINITION_FIELD('Name'),
			required: true,
		},
		{
			...DATA_DEFINITION_FIELD('Email'),
			required: true,
		},
	],
	dataLayoutFields: {
		RequiredName: {required: false},
	},
};

const VIEW_LEVEL_STATE = {
	...STRUCTURE_LEVEL_STATE,
	dataDefinitionField: {
		...DATA_DEFINITION_FIELD('Name'),
		required: false,
	},
	dataLayoutField: {
		...DATA_DEFINITION_FIELD('Name'),
		required: false,
	},
};

const AppContext = createContext();

const RequiredFieldWrapper = ({
	state = STRUCTURE_LEVEL_STATE,
	dataLayoutBuilder = FORM_VIEW.getDataLayoutBuilderProps(),
	dispatch = jest.fn(),
}) => (
	<AppContext.Provider value={[state, jest.fn()]}>
		<RequiredField
			dataLayoutBuilder={dataLayoutBuilder}
			dispatch={dispatch}
			state={state}
		/>
	</AppContext.Provider>
);

describe('RequiredField', () => {
	afterEach(() => {
		cleanup();
		jest.spyOn(utils, 'containsFieldInsideFormBuilder').mockImplementation(
			() => true
		);
	});

	it('renders with structure-level', () => {
		const {asFragment} = render(<RequiredFieldWrapper />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('renders with view-level', () => {
		const {asFragment} = render(
			<RequiredFieldWrapper state={VIEW_LEVEL_STATE} />
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('renders DropDownButton enabled when toggleSwitch was true', () => {
		const {container} = render(<RequiredFieldWrapper />);

		const toggleSwitch = container.querySelector('.toggle-switch-check');

		const dropdownButton = container.querySelector(
			'[title="required-options"]'
		);

		expect(toggleSwitch.checked).toBeTruthy();

		expect(dropdownButton.disabled).toBeFalsy();
	});

	it('renders DropDownButton disabled when toggleSwitch was false', () => {
		const {container} = render(
			<RequiredFieldWrapper state={VIEW_LEVEL_STATE} />
		);

		const toggleSwitch = container.querySelector('.toggle-switch-check');

		const dropdownButton = container.querySelector(
			'[title="required-options"]'
		);

		expect(toggleSwitch.checked).toBeFalsy();

		expect(dropdownButton.disabled).toBeTruthy();
	});

	it('renders the popover with structure level option selected', () => {
		const {container} = render(<RequiredFieldWrapper />);

		const dropdownButton = container.querySelector(
			'[title="required-options"]'
		);
		const popover = container.querySelector(
			'.form-renderer-required-field__popover'
		);
		const options = document.querySelectorAll(
			'.form-renderer-required-field input.custom-control-input'
		);

		expect(popover.classList.contains('show')).toBeFalsy();

		fireEvent.click(dropdownButton);

		expect(popover.classList.contains('show')).toBeTruthy();

		expect(options[0].checked).toBeTruthy();
	});

	it('renders the popover with view level option selected', () => {
		const VIEW_LEVEL_STATE_CUSTUM = {
			...VIEW_LEVEL_STATE,
			dataLayoutField: {
				...DATA_DEFINITION_FIELD('Name'),
				required: true,
			},
		};

		const {container} = render(
			<RequiredFieldWrapper state={VIEW_LEVEL_STATE_CUSTUM} />
		);

		const dropdownButton = container.querySelector(
			'[title="required-options"]'
		);
		const popover = container.querySelector(
			'.form-renderer-required-field__popover'
		);
		const options = document.querySelectorAll(
			'.form-renderer-required-field input.custom-control-input'
		);

		expect(popover.classList.contains('show')).toBeFalsy();

		fireEvent.click(dropdownButton);

		expect(popover.classList.contains('show')).toBeTruthy();

		expect(options[1].checked).toBeTruthy();
	});

	it('change level field', () => {
		const dispatch = jest.fn();

		const {container} = render(
			<RequiredFieldWrapper
				dispatch={dispatch}
				state={VIEW_LEVEL_STATE}
			/>
		);

		const options = container.querySelectorAll(
			'.form-renderer-required-field input.custom-control-input'
		);

		fireEvent.click(options[0]);

		expect(dispatch).toBeCalled();
	});

	it('change ToggleSwitch to false', () => {
		const dataLayoutBuilderProps = FORM_VIEW.getDataLayoutBuilderProps();

		const {container} = render(
			<RequiredFieldWrapper dataLayoutBuilder={dataLayoutBuilderProps} />
		);

		const toggleSwitch = container.querySelector('.toggle-switch-check');

		fireEvent.click(toggleSwitch);

		expect(
			dataLayoutBuilderProps.formBuilderWithLayoutProvider.refs
				.layoutProvider.dispatch
		).toHaveBeenCalledWith('fieldEdited', {
			propertyName: 'required',
			propertyValue: false,
		});
	});

	it('change ToggleSwitch to true', () => {
		const dataLayoutBuilderProps = FORM_VIEW.getDataLayoutBuilderProps();

		const {container} = render(
			<RequiredFieldWrapper
				dataLayoutBuilder={dataLayoutBuilderProps}
				state={VIEW_LEVEL_STATE}
			/>
		);

		const toggleSwitch = container.querySelector('.toggle-switch-check');

		fireEvent.click(toggleSwitch);

		expect(
			dataLayoutBuilderProps.formBuilderWithLayoutProvider.refs
				.layoutProvider.dispatch
		).toHaveBeenCalledWith('fieldEdited', {
			propertyName: 'required',
			propertyValue: true,
		});
	});
});
