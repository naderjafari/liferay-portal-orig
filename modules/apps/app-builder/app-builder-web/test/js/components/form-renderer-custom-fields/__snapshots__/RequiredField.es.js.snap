// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequiredField renders with structure-level 1`] = `
<DocumentFragment>
  <div
    class="form-renderer-required-field d-flex justify-content-between"
  >
    <div
      class="form-group form-renderer-required-field__input"
    >
      <label
        class="toggle-switch simple-toggle-switch"
      >
        <span
          class="toggle-switch-check-bar"
        >
          <input
            checked=""
            class="toggle-switch-check"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="toggle-switch-bar"
          >
            <span
              class="toggle-switch-handle"
            />
          </span>
        </span>
        <span
          class="toggle-switch-label"
        >
          required-field
        </span>
      </label>
    </div>
    <button
      class="form-renderer-required-field__button btn btn-monospaced btn-outline-borderless btn-sm btn-outline-secondary"
      title="required-options"
      type="button"
    >
      <svg
        class="lexicon-icon lexicon-icon-ellipsis-v"
        role="presentation"
      >
        <use
          xlink:href="#ellipsis-v"
        />
      </svg>
    </button>
    <div
      class="form-renderer-required-field__popover popover clay-popover-bottom-right"
    >
      <div
        class="arrow"
      />
      <div
        class="inline-scroller"
        tabindex="-1"
      >
        <div
          class="popover-header"
        >
          required-options
        </div>
        <div
          class="popover-body"
        >
          <div
            class="mt-2"
          >
            <div
              class=""
            >
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    checked=""
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="structure-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      for-all-forms-using-this-field
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="view-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      only-for-this-form
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RequiredField renders with view-level 1`] = `
<DocumentFragment>
  <div
    class="form-renderer-required-field d-flex justify-content-between"
  >
    <div
      class="form-group form-renderer-required-field__input"
    >
      <label
        class="toggle-switch simple-toggle-switch"
      >
        <span
          class="toggle-switch-check-bar"
        >
          <input
            class="toggle-switch-check"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="toggle-switch-bar"
          >
            <span
              class="toggle-switch-handle"
            />
          </span>
        </span>
        <span
          class="toggle-switch-label"
        >
          required-field
        </span>
      </label>
    </div>
    <button
      class="form-renderer-required-field__button btn btn-monospaced btn-outline-borderless btn-sm btn-outline-secondary"
      disabled=""
      title="required-options"
      type="button"
    >
      <svg
        class="lexicon-icon lexicon-icon-ellipsis-v"
        role="presentation"
      >
        <use
          xlink:href="#ellipsis-v"
        />
      </svg>
    </button>
    <div
      class="form-renderer-required-field__popover popover clay-popover-bottom-right"
    >
      <div
        class="arrow"
      />
      <div
        class="inline-scroller"
        tabindex="-1"
      >
        <div
          class="popover-header"
        >
          required-options
        </div>
        <div
          class="popover-body"
        >
          <div
            class="mt-2"
          >
            <div
              class=""
            >
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="structure-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      for-all-forms-using-this-field
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    checked=""
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="view-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      only-for-this-form
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
